sandbox/reflection/struct.h

```diff
@@ -0,0 +1,192 @@
+#ifndef BRICKS_REFLECTION_STRUCT_H
+#define BRICKS_REFLECTION_STRUCT_H
+
+#include <type_traits>
+
+#include "base.h"
+
+namespace bricks {
+namespace reflection {
+
+template <typename INSTANTIATION_TYPE, typename T>
+struct BaseTypeHelperImpl {
+  struct type {};
+};
+
+template <typename T>
+struct BaseTypeHelperImpl<DeclareFields, T> {
+  typedef T type;
+};
+
+template <typename INSTANTIATION_TYPE, typename T>
+using BaseTypeHelper = typename BaseTypeHelperImpl<INSTANTIATION_TYPE, T>::type;
+
+// Fields declaration and counting.
+template <typename INSTANTIATION_TYPE, typename T>
+struct FieldImpl;
+
+template <typename T>
+struct FieldImpl<DeclareFields, T> {
+  typedef T type;
+};
+
+template <typename T>
+struct FieldImpl<CountFields, T> {
+  // TODO: Read on padding.
+  typedef long long type;
+};
+
+template <typename INSTANTIATION_TYPE, typename T>
+using Field = typename FieldImpl<INSTANTIATION_TYPE, T>::type;
+
+// Check if `CURRENT_REFLECTION(f, Index<FieldType, i>)` can be called for i = [0, N].
+template <typename T, int N>
+struct CurrentStructFieldsConsistency {
+  struct Dummy {};
+  constexpr static bool CheckField(char) { return false; }
+  constexpr static auto CheckField(int)
+      -> decltype(T::CURRENT_REFLECTION(Dummy(), Index<FieldType, N>()), bool()) {
+    return true;
+  }
+  constexpr static bool Check() { return CheckField(0) && CurrentStructFieldsConsistency<T, N - 1>::Check(); }
+};
+
+template <typename T>
+struct CurrentStructFieldsConsistency<T, -1> {
+  constexpr static bool Check() { return true; }
+};
+
+}  // namespace reflection
+}  // namespace bricks
+
+#define CURRENT_EXPAND_MACRO_2(x) x
+#define CURRENT_EXPAND_MACRO(x) CURRENT_EXPAND_MACRO_2(x)
+
+// Macros for structure definition.
+#define CURRENT_STRUCT_SWITCH(_1, _2, F, ...) F
+#define CURRENT_STRUCT(...) \
+  CURRENT_STRUCT_SWITCH(__VA_ARGS__, CURRENT_STRUCT_DERIVED, CURRENT_STRUCT_NOT_DERIVED)(__VA_ARGS__)
+
+#define CURRENT_STRUCT_HELPERS(s)                                                        \
+  template <typename INSTANTIATION_TYPE>                                                 \
+  struct CURRENT_STRUCT_IMPL_##s;                                                        \
+  using s = CURRENT_STRUCT_IMPL_##s<bricks::reflection::DeclareFields>;                  \
+  template <typename T>                                                                  \
+  struct CURRENT_REFLECTION_HELPER;                                                      \
+  template <>                                                                            \
+  struct CURRENT_REFLECTION_HELPER<s> {                                                  \
+    constexpr static const char* name() { return #s; }                                   \
+    constexpr static size_t index_base = __COUNTER__;                                    \
+    typedef CURRENT_STRUCT_IMPL_##s<bricks::reflection::CountFields> field_count_struct; \
+  };
```

Comment:
Nit: Best to omit this last semicolon in the expansion of `CURRENT_STRUCT_HELPERS`, and instead put an explicit one on each use.